### PR TITLE
Certora Lockedgold and Accounts verification

### PR DESCRIPTION
**Pull request by lucasege**
### Description

Adding Certora scripts, circleCI config changes, and specifications/harnesses for the LockedGold and Accounts contracts.

### Tested

Changes to contracts are minimal (function visibility settings), but those have been tested.

### Other changes

Changing `Accounts.sol.accounts` and `LockedGold.sol.balances`  to internal.

### Related issues

- #2673 and #2011 - Should not close either issue. 

### Backwards compatibility

Yes

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -845,11 +845,66 @@ jobs:
...
 
+
+  certora-test:
+    working_directory: ~/app
+    docker:
+      - image: circleci/openjdk:11-jdk
+    steps:
+      - attach_workspace:
+          at: ~/app
+      - run:
```
Comment (asaj) [inline .circleci/config.yml:801]: Flagging that we'll want to use the smarter CI running tool @AlexBHarley is working on in https://github.com/celo-org/celo-monorepo/pull/3761

File: .circleci/config.yml [hunk 2]
```diff
@@ -845,11 +845,66 @@ jobs:
...
+            mv solc-static-linux solc
+            chmod +x solc
+      - run:
+          name: Build and run - Locked Gold
+          command: |
+            cd packages/protocol
+            python3 scripts/certora/certoraBuild.py specs/harnesses/GoldTokenHarness.sol specs/harnesses/LockedGoldHarness.sol --link LockedGoldHarness:goldToken=GoldTokenHarness --solc ~/app/solc --verify LockedGoldHarness:specs/lockedGold.spec --settings -assumeUnwindCond
+            CERTORA_RUN_NAME=LockedGoldRun`date +%s`           
+            java -jar scripts/certora/prover_cli.jar verify --name=$CERTORA_RUN_NAME .
+            python3 scripts/certora/getMyCertoraOutput.py celo_ci $CERTORA_RUN_NAME
```
Comment (asaj) [inline .circleci/config.yml]: Can we have this take the output file as an argument?

File: .circleci/config.yml [hunk 3]
```diff
@@ -845,11 +845,66 @@ jobs:
...
+            CERTORA_RUN_NAME=AccountsRun`date +%s`           
+            java -jar scripts/certora/prover_cli.jar verify --name=$CERTORA_RUN_NAME .
+            python3 scripts/certora/getMyCertoraOutput.py celo_ci $CERTORA_RUN_NAME
+            mv out.json AccountsActual.json
+            res2=`python3 scripts/certora/compareResults.py AccountsActual.json specs/AccountsExpected.json`||true
+            rc2=$?
+            echo Result $res2
+            echo Success $rc2
+      - run:
+          name: Check all results are good
```
Comment (asaj) [inline .circleci/config.yml]: Is this redundant? It seems like we do something similar in the `Build and run` steps

File: packages/protocol/accounts.conf [hunk 4]
```diff
@@ -0,0 +1,54 @@
+{
+    "contracts": [
```
Comment (asaj) [inline packages/protocol/accounts.conf]: This appears to have some formatting issues

File: packages/protocol/accounts.conf [hunk 5]
```diff
@@ -0,0 +1,54 @@
+{
```
Comment (asaj) [inline packages/protocol/accounts.conf]: Under what circumstances does this file need to be updated?

File: packages/protocol/lockedGold.conf [hunk 6]
```diff
@@ -0,0 +1,60 @@
+{
```
Comment (asaj) [inline packages/protocol/lockedGold.conf]: Same questions as `account.conf`

File: packages/protocol/accounts.conf (hunk 5 above)
Comment (asaj) [inline packages/protocol/accounts.conf]: Should this file be in a `conf` directory? Seems like we will have one per contract

File: packages/protocol/lockedGold.conf [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        "specs/harnesses/LockedGoldHarness.sol:LockedGoldHarness",
+		"specs/harnesses/GoldTokenHarness.sol"
+		
+    ],
+	"link": {
+        "LockedGoldHarness": {
+            "goldToken": "GoldTokenHarness"
+        }
+    },
+    "output": ".certora_build",
```
Comment (asaj) [inline packages/protocol/lockedGold.conf]: A lot of this looks the same as `accounts.conf`, does it make sense to consolidate the shared pieces?

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 8]
```diff
@@ -0,0 +1,708 @@
...
+import re
+import platform
+from collections import OrderedDict
+from datetime import datetime
+import distutils.dir_util
+import certoraUtils
+from certoraUtils import debug_print, is_windows, safe_create_dir, run_cmd, get_file_basename, get_file_extension, read_from_conf, handle_file_list, get_file_name, current_conf_to_file
+from certoraUtils import OPTION_SOLC, OPTION_SOLC_ARGS, OPTION_SOLC_MAP, OPTION_PATH, OPTION_OUTPUT, OPTION_OUTPUT_FOLDER, OPTION_OUTPUT_VERIFY, OPTION_VERIFY, OPTION_LINK, OPTION_PACKAGES, OPTION_PACKAGES_PATH, OPTION_CACHE, DEFAULT_CONF, OPTION_ADDRESS, OPTION_LINK_CANDIDATES, fatal_error
+
+def print_usage(): # TODO use the macros in print usage as well?
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: I think using `argparse` might be helpful here?

https://docs.python.org/3/library/argparse.html

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 9]
```diff
@@ -0,0 +1,708 @@
...
+	if OPTION_PACKAGES_PATH not in parsed_options:
+		parsed_options[OPTION_PACKAGES_PATH] = os.getenv("NODE_PATH","%s/node_modules" % (os.getcwd())).replace("\\","/")
+		
+	# If packages were not specified, try to find them from package.json, if it exists
+	if OPTION_PACKAGES not in parsed_options:
+		try:
+			with open("package.json","r") as package_json_file:
+				package_json = json.load(package_json_file)
+				deps = set(list(package_json["dependencies"].keys()) if "dependencies" in package_json else list() + list(package_json["devDependencies"].keys()) if "devDependencies" in package_json else list()) # May need both
+				# Don't know which ones we need, so we take them all
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: Remove?

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 10]
```diff
@@ -0,0 +1,708 @@
...
+
+	# Add default for "path"
+	if OPTION_PATH not in parsed_options:
+		parsed_options[OPTION_PATH] = "%s/contracts/" % (os.getcwd().replace("\\","/"))
+		
+	# Add default packages path
+	if OPTION_PACKAGES_PATH not in parsed_options:
+		parsed_options[OPTION_PACKAGES_PATH] = os.getenv("NODE_PATH","%s/node_modules" % (os.getcwd())).replace("\\","/")
+		
+	# If packages were not specified, try to find them from package.json, if it exists
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: Any reason we can't just always read them from package.json? Seems redundant to additionally specify them in the `.conf` file

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+		cygwin_config_path = "/%s" % (config_path[0].lower() + config_path[1:].replace(":",""))
+		print("Cygwin config path: %s" % (cygwin_config_path))
+			
+
+	def collect_funcs(contract,data):	
+		funcs = []
+		abi = data["contracts"][contract]["abi"]
+		for f in filter(lambda x: x["type"]=="function",json.loads(abi)):
+			inputs = f["inputs"]
+			#inputTypes = ",".join(map(lambda x: "\"%s\""%(x["type"]), inputs))
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: Remove?

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 12]
```diff
@@ -0,0 +1,708 @@
...
+			inputTypes = [x["type"] for x in inputs]
+			if "outputs" in f:
+				outputs = f["outputs"]
+				#outputTypes = ",".join(map(lambda x: "\"%s\""%(x["type"]), outputs))
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: Remove?

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 13]
```diff
@@ -0,0 +1,708 @@
...
+	library_addresses = [] # Will become primary contracts too
+		
+	safe_create_dir(certoraUtils.config_path)	
+		
+	if "iscygwin" in parsed_options:
+		cygwin_config_path = "/%s" % (config_path[0].lower() + config_path[1:].replace(":",""))
+		print("Cygwin config path: %s" % (cygwin_config_path))
+			
+
+	def collect_funcs(contract,data):
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: Would `eth-abi` be useful here?

https://pypi.org/project/eth-abi/

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 14]
```diff
@@ -0,0 +1,708 @@
+#!/usr/bin/env python3
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: This file is quite large and hard to review, could it be simplified by using existing libraries? Perhaps made more readable by breaking out some functions here into a different file? I'm afraid I don't fully understand what this is doing.

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 15]
```diff
@@ -0,0 +1,708 @@
...
+			else:
+				isNotPayable = not f["payable"]	# Only if something is definitely non-payable, we treat it as such
+			
+			# Nice to have hex too
+			base = "%s(%s)" % (f["name"],','.join([x.replace('"','') for x in inputTypes]))
+			hash = keccak.new(digest_bits=256)
+			hash.update(str.encode(base))
+			hex = hash.hexdigest()[0:8]
+						
+			funcs.append({"name":f["name"],"args":inputTypes,"returns":outputTypes,"sighash":hex,"notpayable":isNotPayable,"isABI":True})
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: IIUC this just transforms the JSON ABI into a different format, is there any reason the original format can't be used?

File: packages/protocol/scripts/certora/certoraBuild.py [hunk 16]
```diff
@@ -0,0 +1,708 @@
...
+				
+			sources_to_update = get_matching_sdc_names_from_SDCs(src_contract)
+			for source_to_update in sources_to_update:
+				if "state" not in SDCs[source_to_update]:
+					debug_print("Initializing state of %s" %(source_to_update))
+					SDCs[source_to_update]["state"] = {}
+				debug_print("Linking %s (%s) to %s in slot %s" % (src_contract,source_to_update,dst_address,resolved_src_slot))
+				SDCs[source_to_update]["state"][resolved_src_slot] = dst_address
+
+	#if OPTION_ADDRESS in parsed_options:
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraBuild.py]: Remove?

File: packages/protocol/scripts/certora/certoraUtils.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+	return ''.join(file.split("/")[-1])
+
+def get_file_basename(file):
+	return ''.join(file.split("/")[-1].split(".")[0:-1])
+
+def get_file_extension(file):
+	return file.split("/")[-1].split(".")[-1]
+
+def safe_create_dir(path):	
+		try:
```
Comment (asaj) [inline packages/protocol/scripts/certora/certoraUtils.py]: Inconsistent whitespace

File: packages/protocol/specs/AccountsExpected.json [hunk 18]
```diff
@@ -0,0 +1,59 @@
+{
```
Comment (asaj) [inline packages/protocol/specs/AccountsExpected.json]: How is this generated?

File: packages/protocol/scripts/certora/minitester.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+                    gotFuncSet = getFunctionsSet(rulesResults[rule])
+                                        
+                    #warnings = checkMissingFunctions(warnings, testName, rule, expectedFuncSet, gotFuncSet)
+                    
+                    for result, funcList in rulesResults[rule].items():
+                        funcList.sort()
+                        expectedRulesResults[rule][result].sort()
+                        if funcList != expectedRulesResults[rule][result]: # compare functions sets (current results with expected)
+                            for funcName in funcList:
+                                # if funcion appears in current results but does not appear in the expected ones
```
Comment (asaj) [inline packages/protocol/scripts/certora/minitester.py]: typo

File: packages/protocol/scripts/certora/minitester.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+                if type(rulesResults[rule]) == str: # flat rule ( ruleName: result )
+                    if rulesResults[rule] != expectedRulesResults[rule]:
+                        test = False
+                        errors = addError(errors, testName, rule, rulesResults[rule], expectedRulesResults[rule])
+                        
+                else: # nested rule ( ruleName: {result1: [funcionts list], result2: [funcionts list] ... } )
+                    expectedFuncSet = getFunctionsSet(expectedRulesResults[rule])
+                    gotFuncSet = getFunctionsSet(rulesResults[rule])
+                                        
+                    #warnings = checkMissingFunctions(warnings, testName, rule, expectedFuncSet, gotFuncSet)
```
Comment (asaj) [inline packages/protocol/scripts/certora/minitester.py]: remove?

**Review by asaj — COMMENTED** [review]
Still need to review the harnesses

This is exciting! So so so looking forward to having this run in CI 😄 

My main concern is around the python scripts and their possible fragility. It seems like a lot of things are hard coded. Any ideas on how to make them more robust?

File: .circleci/config.yml (hunk 2 above)
Comment (nd-certora) [reply .circleci/config.yml]: Currently, the output.json is a fixed name. Would you like this to be parametric?

File: .circleci/config.yml (hunk 3 above)
Comment (nd-certora) [reply .circleci/config.yml]: We want the verification process to run on all contracts even if one failed. So once everything finished we have another step that checks that everything passed.

File: packages/protocol/accounts.conf (hunk 5 above)
Comment (nd-certora) [reply packages/protocol/accounts.conf]: we are no longer using this file, will remove

File: packages/protocol/specs/AccountsExpected.json (hunk 18 above)
Comment (nd-certora) [reply packages/protocol/specs/AccountsExpected.json]: The content is generated by the tool and we commit this file. This file allows one to indicate that a rule may fail and it is ok. We will change the compare script to not need this file.  In the new version, we will require that everything passes successfully.

File: .circleci/config.yml [hunk 21]
```diff
@@ -862,12 +862,44 @@ jobs:
...
-
+  
+  certora-test:
+    working_directory: ~/app
+    docker:
+      - image: circleci/openjdk:11-jdk
+    steps:
+      - attach_workspace:
+          at: ~/app
+      - run:
```
Comment (asaj) [inline .circleci/config.yml:801]: Per https://github.com/celo-org/celo-monorepo/pull/2721/files#r423128751, can we add the following lines before installing certora dependencies:

```
      - run:
          name: Check if the test should run
          command: |
            ./scripts/ci_check_if_test_should_run_v2.sh @celo/protocol
```

IIUC this should prevent the test from running if nothing has changed in the protocol package or its dependencies

**Review by asaj — APPROVED** [review]
Wow, what an improvement! Go `certora-cli`!! 🤩

File: .circleci/config.yml (hunk 21 above)
Comment (asaj) [reply .circleci/config.yml:801]: Added when resolving conflicts

**Pull request merged**